Makefile

```diff
@@ -0,0 +1,251 @@
+#    and Bob instances represent two traders who can make and take
+#    offers with one another. The Mediator instance represents a Bisq
+#    contributor who can help resolve any technical problems or disputes
+#    that come up between the two traders.
+#
+# STEP 3: Configure the mediator Bisq node. In order to make and take
+# offers, Alice and Bob will need to have a mediator and a refund agent
+# registered on the network. Follow the instructions below to complete
+# that process:
+#
+#  a) Go to the Account screen in the Mediator instance and press CMD+N
+#  and a popup will appear. Click 'Unlock' and then click 'Register' to
+#  register the instance as a mediator.
+#
+#  b) While still in the Account screen, press CMD+D and follow the same
+#  steps as above to register the instance as a refund agent.
+#
+# When the steps above are complete, your localnet should be up and
+# ready to use. You can now test in isolation all Bisq features and use
+# cases.
+#
+
+# Set up everything necessary for deploying your localnet. This is the
+# default target.
+setup: build .localnet
+
+clean: clean-build clean-localnet
+
+clean-build:
+	./gradlew clean
+
+clean-localnet:
+	rm -rf .localnet ./dao-setup
+
+# Build Bisq binaries and shell scripts used in the targets below
+build: seednode/build desktop/build
+
+seednode/build:
+	./gradlew :seednode:build
+
+desktop/build:
+	./gradlew :desktop:build
+
+# Unpack and customize a Bitcoin regtest node and Alice and Bob Bisq
+# nodes that have been preconfigured with a blockchain containing the
+# BSQ genesis transaction
+.localnet:
+	# Unpack the old dao-setup.zip and move things around for more
+	# concise and intuitive naming. This is a temporary measure until we
+	# clean these resources up more thoroughly.
+	unzip docs/dao-setup.zip
+	mv dao-setup .localnet
+	mv .localnet/Bitcoin-regtest .localnet/bitcoind
+	mv .localnet/bisq-BTC_REGTEST_Alice_dao .localnet/alice
+	mv .localnet/bisq-BTC_REGTEST_Bob_dao .localnet/bob
+	# Remove the preconfigured bitcoin.conf in favor of explicitly
+	# parameterizing the invocation of bitcoind in the target below
+	rm -v .localnet/bitcoind/bitcoin.conf
+	# Avoid spurious 'runCommand' errors in the bitcoind log when nc
+	# fails to bind to one of the listed block notification ports
+	echo exit 0 >> .localnet/bitcoind/blocknotify
+
+# Alias '.localnet' to 'localnet' so the target is discoverable in tab
+# completion
+localnet: .localnet
+
+# Deploy a complete localnet by running all required Bitcoin and Bisq
+# nodes, each in their own named screen window. If you are not a screen
+# user, you'll need to manually run each of the targets listed below
+# commands manually in a separate terminal or as background jobs.
+deploy: setup
+	# create a new screen session named 'localnet'
+	screen -dmS localnet
+	# deploy each node in its own named screen window
+	targets=('bitcoind' 'seednode' 'seednode2' 'alice' 'bob' 'mediator'); \
```

Comment:
The Makefile didn't specify the SHELL variable so mine defaulted to /bin/sh which doesn't support the array syntax here. Not sure if there is a shell agnostic way to do it, but adding SHELL=/bin/bash to the Makefile fixed it locally.

```
julian@dev:~/bisq$ make deploy
# create a new screen session named 'localnet'
screen -dmS localnet
# deploy each node in its own named screen window
targets=('bitcoind' 'seednode' 'seednode2' 'alice' 'bob' 'mediator'); \
for t in "${targets[@]}"; do \
	screen -S localnet -X screen -t $t; \
	screen -S localnet -p $t -X stuff "make $t\n"; \
done;
/bin/sh: 1: Syntax error: "(" unexpected
make: *** [Makefile:156: deploy] Error 2

```

Reply:
> The Makefile didn't specify the SHELL variable so mine defaulted to /bin/sh which doesn't support the array syntax here. Not sure if there is a shell agnostic way to do it, but adding SHELL=/bin/bash to the Makefile fixed it locally.

Good catch, thanks @julianknutsen. Commit 234c228df removes the offending bashism such that `/bin/sh` should run without error.

> The iterative single-node behavior also doesn't seem to work quite right. I checked out an older version of the code, killed bob, ran `make bob` and it just restarted without a build.

You're right, and I should have mentioned this in https://github.com/bisq-network/bisq/pull/3718#issuecomment-560573961. The rationale for why things need to be this way is laid out in 5fb4b21. Search for the first occurrence of the word 'contention' there. I'm updating the comment above now.

Reply:
Thanks for the commit link and it makes sense w.r.t. contention. Verified the bashism issue is fixed as well and will do 1.2.4 testing today with your latest code and call out any other glaring issues. So expect a ACK by EOD from me if everything works out.

Reply:
@cbeams thanks for your instructive commit messages!
I understand your reasoning behind removing `build` from PHONY. I'm just wondering wether there is a better solution. Doing a clean-rebuild takes a very long time and is not practical for quick iteration cycles. Essentially the problem you have solved is a race condition. Couldn't we ensure somehow that the build runs just once before deploying the individual nodes. Perhaps running the individual nodes shouldn't depend on the setup/build commands. That way we could have both a 1 time setup + iteration.
Is there a significant advantage of having all the individual node commands depend on setup/build that I'm missing?

Reply:
> Doing a clean-rebuild takes a very long time and is not practical for quick iteration cycles.

Here's what I'm doing in practice as I develop stuff. The following assumes I have already done a `make deploy`, and that now I'm iterating on something, using my `alice` desktop node to see my changes live:
```
# quit my running `alice` node, e.g. with CMD-Q in the UI or ^C on the running process
./gradlew :desktop:build && make alice
```
This picks up and rebuilds just the changes I've made and then deploys those new bits as `alice`.

> Couldn't we ensure somehow that the build runs just once before deploying the individual nodes.

It does effectively run just once. The individual node deployment targets, e.g. `alice` and `bob` depend on `build`. Note that build is, in the latest commits, actually included once again in PHONY, so it does run every time, but that target just depends on the more specific `desktop/build` and `seednode/build` targets, which in turn run only if their respective directories do *not* already exist. The effect here is that when we deploy more than one node or all of them via `make deploy`, the `{desktop|seednode}/build` targets get run once and only once, avoiding the above mentioned race condition and inefficient contention for Gradle resources. If you want to cause a fast (incremental) rebuild, it's simply necessary to drop down to calling gradle directly like I've shown above. calling `make clean-build` should not be necessary in any case, unless you actually want to blow away all the build directories.

> Is there a significant advantage of having all the individual node commands depend on setup/build that I'm missing?

It just ensures that deploying any given node causes `build` and `localnet` to run if they have not already done so. So someone can come along in a clean checkout and run only `make bitcoind` and `make seednode` and everything will work as they expect, meaning that the .localnet dir will get created and the seednode build will run. If that's already happened, then those targets are no-ops.